From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: okx-code <okx@okx.sh>
Date: Mon, 21 Jul 2025 22:41:36 +0100
Subject: [PATCH] Optimise mob despawning


diff --git a/net/minecraft/server/level/ServerLevel.java b/net/minecraft/server/level/ServerLevel.java
index 0651a536ad38c3c6d1ac2423b6d351738719339f..7ae929d9d3e494189f89c13c0795df07db64ab98 100644
--- a/net/minecraft/server/level/ServerLevel.java
+++ b/net/minecraft/server/level/ServerLevel.java
@@ -795,6 +795,7 @@ public class ServerLevel extends Level implements ServerEntityGetter, WorldGenLe
             }
 
             io.papermc.paper.entity.activation.ActivationRange.activateEntities(this); // Paper - EAR
+            generateEligiblePlayersForDespawning();
             this.entityTickList
                 .forEach(
                     entity -> {
@@ -2748,4 +2749,24 @@ public class ServerLevel extends Level implements ServerEntityGetter, WorldGenLe
         this.lagCompensationTick = (System.nanoTime() - MinecraftServer.SERVER_INIT) / (java.util.concurrent.TimeUnit.MILLISECONDS.toNanos(50L));
     }
     // Paper end - lag compensation
+
+    // Paper start - optimise despawns
+    @Nullable
+    private io.papermc.paper.util.kdtree.Kd3Tree despawnEligiblePlayers;
+
+    private void generateEligiblePlayersForDespawning() {
+        List<Vec3> nodes = new ArrayList<>();
+        for (ServerPlayer player : players) {
+            if (net.minecraft.world.entity.EntitySelector.PLAYER_AFFECTS_SPAWNING.test(player)) {
+                nodes.add(player.position());
+            }
+        }
+        this.despawnEligiblePlayers = new io.papermc.paper.util.kdtree.Kd3Tree(nodes.toArray(new Vec3[0]));
+    }
+
+    @Nullable
+    public io.papermc.paper.util.kdtree.Kd3Tree getDespawnEligiblePlayers() {
+        return despawnEligiblePlayers;
+    }
+    // Paper end - optimise despawns
 }
diff --git a/net/minecraft/world/entity/Mob.java b/net/minecraft/world/entity/Mob.java
index e330bf990e4874baed1b21cd8c9b44d66ec5b823..85b0af8c521979e84214c6ede5c175b48ec2a6cc 100644
--- a/net/minecraft/world/entity/Mob.java
+++ b/net/minecraft/world/entity/Mob.java
@@ -813,15 +813,17 @@ public abstract class Mob extends LivingEntity implements EquipmentUser, Leashab
         if (this.level().getDifficulty() == Difficulty.PEACEFUL && this.shouldDespawnInPeaceful()) {
             this.discard(EntityRemoveEvent.Cause.DESPAWN); // CraftBukkit - add Bukkit remove cause
         } else if (!this.isPersistenceRequired() && !this.requiresCustomPersistence()) {
-            Entity nearestPlayer = this.level().findNearbyPlayer(this, -1.0, EntitySelector.PLAYER_AFFECTS_SPAWNING); // Paper - Affects Spawning API
-            if (nearestPlayer != null) {
+
+            final io.papermc.paper.util.kdtree.Kd3Tree despawnEligiblePlayers = ((ServerLevel) this.level()).getDespawnEligiblePlayers();
+            if (despawnEligiblePlayers != null && !despawnEligiblePlayers.isEmpty()) {
+                final net.minecraft.world.phys.Vec3 nearest = despawnEligiblePlayers.nearest(this.position());
                 // Paper start - Configurable despawn distances
                 final io.papermc.paper.configuration.WorldConfiguration.Entities.Spawning.DespawnRangePair despawnRangePair = this.level().paperConfig().entities.spawning.despawnRanges.get(this.getType().getCategory());
                 final io.papermc.paper.configuration.type.DespawnRange.Shape shape = this.level().paperConfig().entities.spawning.despawnRangeShape;
-                final double dy = Math.abs(nearestPlayer.getY() - this.getY());
+                final double dy = Math.abs(nearest.y() - this.getY());
                 final double dySqr = Math.pow(dy, 2);
-                final double dxSqr = Math.pow(nearestPlayer.getX() - this.getX(), 2);
-                final double dzSqr = Math.pow(nearestPlayer.getZ() - this.getZ(), 2);
+                final double dxSqr = Math.pow(nearest.x() - this.getX(), 2);
+                final double dzSqr = Math.pow(nearest.z() - this.getZ(), 2);
                 final double distanceSquared = dxSqr + dzSqr + dySqr;
                 // Despawn if hard/soft limit is exceeded
                 if (despawnRangePair.hard().shouldDespawn(shape, dxSqr, dySqr, dzSqr, dy) && this.removeWhenFarAway(distanceSquared)) {
